src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -971,31 +954,57 @@ private void AddEasyHandle(RequestCompletionSource state)
             }
         }
 
+        private static void RemoveEasyHandle(SafeCurlMultiHandle multiHandle, GCHandle stateHandle, RequestCompletionSource state)
+        {
+            RemoveEasyHandle(multiHandle, state, false);
+            if (stateHandle.IsAllocated)
+            {
+                stateHandle.Free();
+            }
+        }
+
         private static void RemoveEasyHandle(SafeCurlMultiHandle multiHandle, GCHandle stateHandle, bool onMultiStack)
         {
+            Debug.Assert(stateHandle.IsAllocated);
+
             RequestCompletionSource state = (RequestCompletionSource)stateHandle.Target;
+            RemoveEasyHandle(multiHandle, state, onMultiStack);
+            stateHandle.Free();
```

Comment:
I'd suggest moving the stateHandle.Free() up one line, to be as soon as you're done with it.